File: leetcode/116-populating-next-right-pointers-in-each-node.py
# Definition for binary tree with next pointer.
# class TreeLinkNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution(object):
    def connect(self, root):
        """
        :type root: TreeLinkNode
        :rtype: nothing
        """
        i = root
        while i:
            if i.left is None:
                break
            j = i
            while j:
                j.left.next = j.right
                if j.next is not None:
                    j.right.next = j.next.left
                j = j.next
            i = i.left

File: leetcode/116-populating-next-right-pointers-in-each-node.cpp
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *     int val;
 *     TreeLinkNode *left, *right, *next;
 *     TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
 public:
  void connect(TreeLinkNode* root) {
    for (TreeLinkNode* i = root; i; i = i->left) {
      if (!i->left) break;
      for (TreeLinkNode *j = i; j; j = j->next) {
        j->left->next = j->right;
        if (j->next)
          j->right->next = j->next->left;
      }
    }
  }
};
